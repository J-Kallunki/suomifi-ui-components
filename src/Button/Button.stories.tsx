import React from 'react';
import { storiesOf } from '@storybook/react';
declare module '@storybook/react' {
  export interface Story {
    addWithJSX(storyName: string, callback: RenderFunction): this;
  }
}

import Button from './Button';
import { withInfo } from '@storybook/addon-info';

storiesOf('Components/Button', module).addWithJSX(
  'basic Button',
  withInfo({
    inline: true,
    source: false,
    text: `

  ### Notes

  This is a button

  ### Usage
  ~~~js
  <Button>Hello</Button>
  ~~~`,
  })(() => <Button>Button</Button>),
);
