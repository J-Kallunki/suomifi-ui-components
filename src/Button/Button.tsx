import React, { Component, ReactNode } from 'react';

export interface ButtonProps {
  children?: ReactNode;
}

export default class Button extends Component<ButtonProps> {
  render() {
    const { children } = this.props;
    // DUMMY TEST Button-component
    // TODO: Replace with real component
    return <button data-testid="button">{children}</button>;
  }
}
