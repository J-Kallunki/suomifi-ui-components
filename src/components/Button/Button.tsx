import React, { Component, MouseEvent, ReactNode } from 'react';
import PropTypes from 'prop-types';

export interface IButtonProps {
  /** Custom class name to append to button */
  className?: string;
  /** Event handler to execute when clicked
   *  @default void
   */
  onClick?: (event: MouseEvent) => void;
  /** data-testid
   *  @default test
   */
  testId?: string;
  children?: ReactNode;
}

export default class Button extends Component<IButtonProps> {
  static defaultProps = {
    testId: 'button',
  };

  static propTypes = {
    className: PropTypes.string,
    onClick: PropTypes.func,
    testId: PropTypes.string,
  };

  render() {
    const { testId, children, ...passProps } = this.props;
    return (
      <button {...passProps} data-testid={testId}>
        {children}
      </button>
    );
  }
}
