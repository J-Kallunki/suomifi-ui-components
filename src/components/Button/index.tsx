export { default as Button } from './Button';
export { IButtonProps } from './Button';
