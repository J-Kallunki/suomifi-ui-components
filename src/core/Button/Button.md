```js
<Button
  className="pizza-button"
  onClick={() => console.log('Gimme pizza!')}
  testId="pizza-test"
>
  🍕
</Button>
```
