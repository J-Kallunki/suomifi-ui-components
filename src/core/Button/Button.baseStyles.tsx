import { css } from '@emotion/core';
import { suomifiTheme, ThemeProp } from '../theme';
import { ButtonProps } from './Button';
import { element, focus, fonts } from '../theme/reset';

const fullWidthStyles = css`
  display: block;
  width: 100%;
`;

const negativeStyles = (theme: ThemeProp) => css`
  background: none;
  background-color: ${theme.colors.highlightBase};
  border: 1px solid ${theme.colors.whiteBase};
  text-shadow: none;

  &:hover {
    background: ${theme.gradients.whiteBaseNegative};
  }

  &:active {
    background: none;
    background-color: ${theme.colors.highlightBase};
  }

  &[disabled],
  &:disabled {
    opacity: 0.5;
    background: none;
  }
`;

const secondaryStyles = (theme: ThemeProp) => css`
  color: ${theme.colors.highlightBase};
  background: none;
  background-color: ${theme.colors.whiteBase};
  border: 1px solid ${theme.colors.highlightBase};
  text-shadow: none;

  &:hover {
    background: ${theme.gradients.highlightLight45};
  }

  &:active {
    background: none;
    background-color: ${theme.colors.highlightLight53};
  }

  &[disabled],
  &:disabled {
    color: ${theme.colors.depthBase};
    text-shadow: none;
    background: none;
    background-color: ${theme.colors.highlightLight53};
    border-color: ${theme.colors.depthBase};
  }
`;

const secondaryNoBorderStyles = (theme: ThemeProp) => css`
  ${secondaryStyles(theme)}
  border: none;
`;

const tertiaryStyles = (theme: ThemeProp) => css`
  ${secondaryStyles(theme)}
  background: ${theme.colors.highlightLight50};
  border: none;

  &:hover {
    background: ${theme.colors.highlightLight53};
  }

  &:active {
    background: ${theme.colors.highlightLight53};
  }
`;

export const baseStyles = ({
  theme = suomifiTheme,
  fullWidth = false,
  variant = 'default',
}: ButtonProps) => css`
  ${element}
  ${fonts.inputSemibold}
  ${focus}
  padding: 10px 20px;
  min-height: 40px;
  color: ${theme.colors.whiteBase};
  background: ${theme.gradients.highlightBase};
  border-radius: ${theme.radius.basic};
  text-align: center;
  text-shadow: ${theme.shadows.invertTextShadow};
  cursor: pointer;

  &:hover {
    background: ${theme.gradients.highlightLight4};
  }

  &:active {
    background: ${theme.gradients.highlightDark9};
  }

  &[disabled],
  &:disabled {
    text-shadow: 0 1px 1px ${theme.colors.blackBase};
    background: ${theme.gradients.depthBase};
    pointer-events: none;
    user-select: none;
  }

  ${fullWidth && fullWidthStyles}
  ${variant === 'negative' && negativeStyles(theme)}
  ${variant === 'secondary' && secondaryStyles(theme)}
  ${variant === 'secondary-noborder' && secondaryNoBorderStyles(theme)}
  ${variant === 'tertiary' && tertiaryStyles(theme)}

  & > .fi-button-icon {
    width: 16px;
    height: 16px;
    margin-right: 8px;
    &.fi-button-icon--right {
      margin-right: 0;
      margin-left: 8px;
    }
  }
`;

export const unStyled = ({ theme = suomifiTheme }) => css`
  ${element}
  ${focus}
  border-radius: ${theme.radius.basic};
  cursor: pointer;
`;
