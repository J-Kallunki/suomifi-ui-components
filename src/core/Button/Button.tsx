import React, { Component } from 'react';
import styled from '@emotion/styled';
import classnames from 'classnames';
import { ThemeComponent, ThemeProp } from '../theme';
import { withDefaultTheme } from '../theme/utils/defaultTheme';
import { baseStyles, iconBaseStyles } from './Button.baseStyles';
import {
  Button as CompButton,
  ButtonProps as CompButtonProps,
} from '../../components/Button/Button';
import { Icon, IconProps, IconKeys } from '../Icon/Icon';

type ButtonVariant =
  | 'default'
  | 'negative'
  | 'secondary'
  | 'secondary-noborder'
  | 'tertiary'
  | 'unstyled';

export interface ButtonProps extends CompButtonProps, ThemeComponent {
  /**
   * Set width to grow all available space
   */
  fullWidth?: boolean;
  /**
   * 'default' | 'negative' | 'secondary' | 'secondary-noborder' | 'tertiary'
   * @default default
   */
  variant?: ButtonVariant;
  /**
   * Icon from suomifi-theme
   */
  icon?: IconKeys;
  /**
   * Icon from suomifi-theme to be placed on right side
   */
  iconRight?: IconKeys;
  /**
   * Properties given to Icon-component
   */
  iconProps?: IconProps;
}

const baseClassName = 'fi-button';

const StyledButton = styled(
  ({
    theme,
    fullWidth,
    variant,
    className,
    ...passProps
  }: ButtonProps & { right?: boolean }) => (
    <CompButton
      {...passProps}
      className={classnames(className, baseClassName)}
    />
  ),
)`
  label: ${({ disabled }) =>
    disabled ? `${baseClassName}--disabled` : baseClassName};
  ${props => baseStyles(props)}
`;

const StyledIcon = styled(
  ({ right, ...passProps }: IconProps & { right?: boolean }) => (
    <Icon {...passProps} />
  ),
)`
  ${props => iconBaseStyles(props)}
`;

const iconColor = ({
  theme,
  invert,
  disabled,
}: {
  theme?: ThemeProp;
  invert?: boolean;
  disabled?: boolean;
}) => {
  if (!!theme) {
    const defaultColor = !!disabled
      ? theme.colors.disabledColor
      : theme.colors.white;
    const secondaryColor = !!disabled
      ? theme.colors.disabledColor
      : theme.colors.secondaryColor;
    return !!invert ? secondaryColor : defaultColor;
  }
  return undefined;
};

const UnstyledButton = (props: ButtonProps) => {
  const {
    className,
    theme: dissmissTheme,
    fullWidth: dissmissFullWidth,
    variant: dissmissVariant,
    icon: dismissIcon,
    iconRight: dissmissIconRight,
    ...passProps
  } = props;
  return (
    <CompButton
      {...passProps}
      className={classnames(className, baseClassName)}
    />
  );
};

class ButtonWithIcon extends Component<ButtonProps> {
  render() {
    const {
      children,
      icon,
      iconRight,
      iconProps,
      ...passProps
    } = withDefaultTheme(this.props);
    const { theme, disabled, variant } = passProps;

    if (variant === 'unstyled') {
      return <UnstyledButton {...passProps} />;
    }

    const secondaryOrTertiary =
      !!variant &&
      (variant === 'secondary' ||
        variant === 'secondary-noborder' ||
        variant === 'tertiary');
    return (
      <StyledButton {...passProps}>
        {!!icon && (
          <StyledIcon
            icon={icon}
            right={false}
            color={iconColor({ theme, disabled, invert: secondaryOrTertiary })}
            {...iconProps}
          />
        )}
        {children}
        {!!iconRight && (
          <StyledIcon
            icon={iconRight}
            right={true}
            color={iconColor({ theme, disabled, invert: secondaryOrTertiary })}
            {...iconProps}
          />
        )}
      </StyledButton>
    );
  }
}

/**
 * Use for inside Application onClick events.<br />
 * When using Button.secondaryNoborder with other than white background,<br />
 * define styles background color for all needed states (:hover, :active, :disabled)
 */
export class Button extends Component<ButtonProps> {
  static negative = (props: ButtonProps) => {
    return <ButtonWithIcon {...props} variant="negative" />;
  };

  static secondary = (props: ButtonProps) => {
    return <ButtonWithIcon {...props} variant="secondary" />;
  };

  static secondaryNoborder = (props: ButtonProps) => {
    return <ButtonWithIcon {...props} variant="secondary-noborder" />;
  };

  static tertiary = (props: ButtonProps) => {
    return <ButtonWithIcon {...props} variant="tertiary" />;
  };

  static unstyled = (props: ButtonProps) => {
    return <UnstyledButton {...props} />;
  };

  render() {
    return <ButtonWithIcon {...this.props} />;
  }
}
