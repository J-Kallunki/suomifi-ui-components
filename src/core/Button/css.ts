import { cssFromBaseStyles } from '../utils';
import { baseStyles } from './Button.baseStyles';

export const css = cssFromBaseStyles(baseStyles);
