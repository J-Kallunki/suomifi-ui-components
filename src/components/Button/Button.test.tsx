import React from 'react';
import { render } from 'react-testing-library';

import Button from './Button';

test('calling render with the same component on the same container does not remount', () => {
  const buttonRendered = render(<Button>Test</Button>);
  const { getByTestId, rerender } = buttonRendered;
  expect(getByTestId('button').textContent).toBe('Test');

  // re-render the same component with different props
  rerender(<Button>Test two</Button>);
  expect(getByTestId('button').textContent).toBe('Test two');
});
