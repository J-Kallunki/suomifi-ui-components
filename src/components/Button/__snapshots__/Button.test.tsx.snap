// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.emotion-0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  line-height: 1;
  text-align: left;
  vertical-align: baseline;
  background: none;
  display: inline-block;
  max-width: 100%;
  pointer: cursor;
}

.emotion-0:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.emotion-0::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.emotion-0::-webkit-inner-spin-button {
  height: auto;
}

.emotion-0::-webkit-outer-spin-button {
  height: auto;
}

<button
  aria-disabled="false"
  class="emotion-0"
  data-testid="button"
  tabindex="0"
  type="button"
>
  Test
</button>
`;
